janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/MultiPageEntryBuffer.java

```diff
@@ -0,0 +1,399 @@
+                int addLimit;
+                int delLimit;
+                if (nextPageStart == null) //this is the last page, everything we still need to add/delete applies to it
+                {
+                    addLimit = add.length;
+                    delLimit = del.length;
+                } else //this is not the last page, we need to determine which adds/deletes go to this page, and which go to next page(s)
+                {
+                    //NOTE: for long mutation lists, it could be better to do binary search here,
+                    // otherwise it could be up to maxPageSize linear comparisons.
+                    // However it was not seen as a bottleneck in practice so far
+                    addLimit = iadd;
+                    while (addLimit < add.length && nextPageStart.compareTo(add[addLimit]) > 0) {
+                        addLimit++;
+                    }
+
+                    delLimit = idel;
+                    while (delLimit < del.length && nextPageStart.compareTo(del[delLimit]) > 0) {
+                        delLimit++;
+                    }
+                }
+
+                List<BufferPage> mergedPages = currPage.merge(add, iadd, addLimit, del, idel, delLimit, maxPageSize);
+                if (mergedPages.size() == 0) //there was no data left in the page as a result of merge - remove old page
+                {
+                    pages.remove(currPageNo);
+                    //do NOT increase currPageNo here as the next page moved in to this place
+                } else //there is at least one page as a result of merge - replace the current one and insert any additional overflow pages
+                {
+                    pages.set(currPageNo, mergedPages.get(0)); //replace the currPage with the newly merged version
+                    currPageNo++; //move to next page
+                    if (mergedPages.size() > 1) //more than one page as a result of merge - insert all additional ones
+                    {
+                        mergedPages.remove(0);
+                        pages.addAll(currPageNo, mergedPages);
+
+                        //skip over the pages we just added as they cannot contain any work we might still need to do
+                        currPageNo += mergedPages.size();
+                        pageHits += mergedPages.size();
+                    }
+                }
+
+                iadd = addLimit;
+                idel = delLimit;
+
+                pageHits++;
+            } else {
+                currPageNo++;
+            }
+        }
+
+        if (oldPageCount >= pages.size()) {
+            //it grew before but not this time, assume it stopped growing for now and trim to size to save memory
+            pages.trimToSize();
+        }
+    }
+
+    @Override
+    public boolean isPaged() {
+        return true;
+    }
+
+    @Override
+    public FragmentationReport createFragmentationReport(int maxPageSize) {
```

Comment:
To create FragmentationReport, it would be good to use a builder.

Reply:
done, although it is not intended to ever be constructed by user code, and basically is likely to only be ever constructed in one place, so not so much scope for using this builder...